### ISSUE TITLE
Dynamic/conditional dependencies

### ISSUE THREAD

**Issue by gbj** [human-solutions/reactive-signals#2]
Really cool work here, especially on the data structures and the performance they enable.

I thought I'd open a friendly issue to add something to your TBD/TBCs, unless I missed it.

From the readme I see you already have the "diamond problem" (`1 kiwi, 1 kiwi`) on your radar. Just wanted to raise as well the question of dynamic/conditional dependencies.

For example my favorite "full name" example... The "effect" here (i.e., the signal that runs a relatively expensive operating, printing out to the console) runs unnecessarily several times because it never unsubscribes from `last_name`, even in the branch in which it's not used. 

Dynamic dependencies can be a great optimization, because even though they slow down signal updates slightly, they make sure that you are running the effect the fewest times possible... and in most use cases running the effects is several orders of magnitude more expensive than updating signals, so it's often a worthwhile tradeoff. 

See the duplicate prints of `Bill` in this example when I change `last` even though `use_last` is set to `false`. These can be a single `Bill`. 

```rust
fn main() {
    use reactive_signals::{runtimes::ClientRuntime, signal};

    // signals are created in scopes
    let cx = ClientRuntime::new_root_scope();
    let first = signal!(cx, "Greg");
    let last = signal!(cx, "Johnston");
    let use_last = signal!(cx, true);

    // is_plural will update when count changes
    let full_name = signal!(cx, move || if use_last.get() {
        format!("{} {}", first.get(), last.get())
    } else {
        first.get().to_string()
    });

    signal!(cx, move || {
        println!("{}", full_name.cloned());
    });

    first.set("Bill");
    last.set("Thompson");
    use_last.set(false);
    last.set("Jones");
    last.set("Smith");
    use_last.set(true);
}
```

**Comment by akesson** [comment]
Yes, you are right. I'll add it to the TBCs. The unsubscribes are something that would make sense to solve at the same time as the diamond problem.

But I'll hold off on any improvements until I have a better understanding of how `reactive-signals` is received.
